diff --git a/src/core/bspline.cpp b/src/core/bspline.cpp
index 9ac3a48..3e06d7c 100644
--- a/src/core/bspline.cpp
+++ b/src/core/bspline.cpp
@@ -206,4 +206,158 @@ splineeval(const double *knots, const double *weights, int nknots, double x, int
 //	return cblas_sdot(totalcoeff, basis1->data, 1, basis2->data, 1);
 //}
 	
+void
+bsplvb_simple(const double *knots, const unsigned nknots,
+    double x, int left, int degree, float* biatx)
+{
+	int i, j;
+	double saved, term;
+	double delta_l[degree], delta_r[degree];
+	
+	biatx[0] = 1.0;
+	
+	/*
+	 * Handle the (rare) cases where x is outside the full
+	 * support of the spline surface.
+	 */
+	if (left == degree-1)
+		while (left >= 0 && x < knots[left])
+			left--;
+	else if (left == nknots-degree-1)
+		while (left < nknots-1 && x > knots[left+1])
+			left++;	
+	
+	/* 
+	 * NB: if left < degree-1 or left > nknots-degree-1,
+	 * the following loop will dereference addresses ouside
+	 * of knots[0:nknots]. While terms involving invalid knot
+	 * indices will be discarded, it is important that `knots'
+	 * have (maxdegree-1)*sizeof(double) bytes of padding
+	 * before and after its valid range to prevent segfaults
+	 * (see parsefitstable()).
+	 */
+	for (j = 0; j < degree-1; j++) {
+		delta_r[j] = knots[left+j+1] - x;
+		delta_l[j] = x - knots[left-j];
+		
+		saved = 0.0;
+		
+		for (i = 0; i < j+1; i++) {
+			term = biatx[i] / (delta_r[i] + delta_l[j-i]);
+			biatx[i] = saved + delta_r[i]*term;
+			saved = delta_l[j-i]*term;
+		}
+		
+		biatx[j+1] = saved;
+	}
+	
+	/* 
+	 * If left < (spline order), only the first (left+1)
+	 * splines are valid; the remainder are utter nonsense.
+	 */
+	if ((i = degree-1-left) > 0) {
+		for (j = 0; j < left+1; j++)
+			biatx[j] = biatx[j+i]; /* Move valid splines over. */
+		for ( ; j < degree; j++)
+			biatx[j] = 0.0; /* The rest are zero by construction. */
+	} else if ((i = left+degree+1-nknots) > 0) {
+		for (j = degree-1; j > i-1; j--)
+			biatx[j] = biatx[j-i];
+		for ( ; j >= 0; j--)
+			biatx[j] = 0.0;
+	}
+}
+
+void
+bsplvb(const double* knots, const double x, const int left, const int jlow,
+    const int jhigh, float* biatx, double* delta_l, double* delta_r)
+{
+	int i, j;
+	double saved, term;
+
+	if (jlow == 0)
+		biatx[0] = 1.0;
+		
+	for (j = jlow; j < jhigh-1; j++) {
+		delta_r[j] = knots[left+j+1] - x;
+		delta_l[j] = x - knots[left-j];
+		
+		saved = 0.0;
+		
+		for (i = 0; i < j+1; i++) {
+			term = biatx[i] / (delta_r[i] + delta_l[j-i]);
+			biatx[i] = saved + delta_r[i]*term;
+			saved = delta_l[j-i]*term;
+		}
+		
+		biatx[j+1] = saved;
+	}
+}
+
+
+void
+bspline_deriv_nonzero(const double* knots, const unsigned nknots,
+    const double x, int left, const int n, float* biatx)
+{
+	int i, j;
+	double temp, a;
+	double delta_l[n], delta_r[n];
+	
+	/* Special case for constant splines */
+	if (n == 0)
+		return;
+	
+	/*
+	 * Handle the (rare) cases where x is outside the full
+	 * support of the spline surface.
+	 */
+	if (left == n)
+		while (left >= 0 && x < knots[left])
+			left--;
+	else if (left == nknots-n-2)
+		while (left < nknots-1 && x > knots[left+1])
+			left++;
+	
+	/* Get the non-zero n-1th order B-splines at x */
+	bsplvb(knots, x, left, 0 /* jlow */, n /* jhigh */,
+	    biatx, delta_l, delta_r);
+	
+	/* 
+	 * Now, form the derivatives of the nth order B-splines from
+	 * linear combinations of the lower-order splines.
+	 */
+	
+	/* 
+	 * On the last supported segment of the ith nth order spline,
+	 * only the i+1th n-1th order spline is nonzero.
+	 */
+	temp = biatx[0];
+	biatx[0] =  - n*temp / ((knots[left+1] - knots[left+1-n]));
+	
+	/* On the middle segments, both the ith and i+1th splines contribute. */
+	for (i = 1; i < n; i++) {
+		a = n*temp/((knots[left+i] - knots[left+i-n]));
+		temp = biatx[i];
+		biatx[i] = a - n*temp/(knots[left+i+1] - knots[left+i+1-n]);
+	}
+	/*
+	 * On the first supported segment of the i+nth nth order spline,
+	 * only the ith n-1th order spline is nonzero.
+	 */
+	biatx[n] = n*temp/((knots[left+n] - knots[left]));
+
+	/* Rearrange for partially-supported points. */
+	if ((i = n-left) > 0) {
+		for (j = 0; j < left+1; j++)
+			biatx[j] = biatx[j+i]; /* Move valid splines over. */
+		for ( ; j < n+1; j++)
+			biatx[j] = 0.0; /* The rest are zero by construction. */
+	} else if ((i = left+n+2-nknots) > 0) {
+		for (j = n; j > i-1; j--)
+			biatx[j] = biatx[j-i];
+		for ( ; j >= 0; j--)
+			biatx[j] = 0.0;
+	}
+
+}
 } //namespace photospline
