diff --git a/include/photospline/detail/splineutil.h b/include/photospline/detail/splineutil.h
index bd7b985..828cbdc 100644
--- a/include/photospline/detail/splineutil.h
+++ b/include/photospline/detail/splineutil.h
@@ -2,12 +2,12 @@
 #ifndef PHOTOSPLINE_CFITTER_SPLINEUTIL_H
 #define PHOTOSPLINE_CFITTER_SPLINEUTIL_H
 
+#include <cholmod.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#include <cholmod.h>
-
 struct ndsparse {
 	/*
 	 * This is an ntuple, and is similar to the CHOLMOD triplet
